typing/ikinds/ldd_hybrid_opt.ml

```diff
@@ -0,0 +1,600 @@
+      | Some vars -> (
+        match
+          List.find_opt
+            (fun (v : t) ->
+              match v.state with
+              | Rigid name' -> name == name' || V.compare name name' = 0
+              | Unsolved | Solved _ -> false)
+            vars
+        with
+        | Some v -> v
+        | None ->
+          let v = { id; state = Rigid name; var_node = bot } in
+          v.var_node <- node_raw v bot top;
+          Hashtbl.replace rigid_tbl id (v :: vars);
+          v)
+  end
+
+  (* Subtract subsets h - l (no memoization). *)
+  let rec canonicalize (h : node) (l : node) : node =
+    if h == l
+    then bot
+    else if l == bot
+    then h
+    else if is_leaf h
+    then
+      let dh = leaf_value h in
+      let dl = down0 l in
+      if C.equal dl C.bot then h else leaf (C.co_sub dh dl)
+    else if is_leaf l
+    then canonicalize_right_leaf h l
+    else
+      let hb = node_block h in
+      let lb = node_block l in
+      let cmp = compare_var hb.v lb.v in
+      if cmp = 0
+      then
+        let lo' = canonicalize hb.lo lb.lo in
+        let hi1 = if lb.hi == bot then hb.hi else canonicalize hb.hi lb.hi in
+        let hi' = if lb.lo == bot then hi1 else canonicalize hi1 lb.lo in
+        if hb.lo == lo' && hb.hi == hi' then h else node_raw hb.v lo' hi'
+      else if cmp < 0
+      then
+        let lo' = canonicalize hb.lo l in
+        let hi' = canonicalize hb.hi l in
+        if hb.lo == lo' && hb.hi == hi' then h else node_raw hb.v lo' hi'
+      else canonicalize h lb.lo
+
+  and canonicalize_right_leaf (h : node) (leaf_l : node) : node =
+    let leaf_val = leaf_value leaf_l in
+    if C.equal leaf_val C.bot then h else canonicalize_right_leaf_aux h leaf_val
+
+  and canonicalize_right_leaf_aux (h : node) (leaf_l : C.t) : node =
+    if is_leaf h
+    then leaf (C.co_sub (leaf_value h) leaf_l)
+    else
+      let hb = node_block h in
+      let lo' = canonicalize_right_leaf_aux hb.lo leaf_l in
+      let hi' = canonicalize_right_leaf_aux hb.hi leaf_l in
+      if hb.lo == lo' && hb.hi == hi' then h else node_raw hb.v lo' hi'
+
+  let node (v : var) (lo : node) (hi : node) : node =
+    if lo == bot then node_raw v lo hi else node_raw v lo (canonicalize hi lo)
+
+  (* --------- boolean algebra over nodes (no memoization) --------- *)
+  let rec join (a : node) (b : node) =
+    if a == b
+    then a
+    else if is_leaf a
+    then join_with_leaf a b
+    else if is_leaf b
+    then join_with_leaf b a
+    else
+      let na = node_block a in
+      let nb = node_block b in
+      let cmp = compare_var na.v nb.v in
+      if cmp = 0
+      then
+        node_raw na.v (join na.lo nb.lo)
+          (join (canonicalize na.hi nb.lo) (canonicalize nb.hi na.lo))
+      else if cmp < 0
+      then node_raw na.v (join na.lo b) (canonicalize na.hi b)
+      else node_raw nb.v (join a nb.lo) (canonicalize nb.hi a)
+
+  and join_with_leaf (leaf_a : node) (other : node) =
+    let leaf_val = leaf_value leaf_a in
```

Comment:
It would make me feel better if functions like this take in a `C.t` instead and the call to `leaf_value` was moved to immediately after the `is_leaf` test. I think this would make it harder to make a mistake. As is, the type of this function is a bit of a lie since it implicitly assumes that `leaf_a` is a leaf.